src/objective/rank_obj.cu

```diff
@@ -0,0 +1,637 @@
+                                 std::vector<MAPStats> *p_map_acc) {
+    std::vector<MAPStats> &map_acc = *p_map_acc;
+    map_acc.resize(sorted_list.size());
+    bst_float hit = 0, acc1 = 0, acc2 = 0, acc3 = 0;
+    for (size_t i = 1; i <= sorted_list.size(); ++i) {
+      if (sorted_list[i - 1].label > 0.0f) {
+        hit++;
+        acc1 += hit / i;
+        acc2 += (hit - 1) / i;
+        acc3 += (hit + 1) / i;
+      }
+      map_acc[i - 1] = MAPStats(acc1, acc2, acc3, hit);
+    }
+  }
+
+ public:
+  // Stopgap method - will be removed when we support other type of ranking - ndcg, map etc.
+  // on GPU later
+  inline static bool SupportOnGPU() { return false; }
+
+  static void GetLambdaWeight(const std::vector<ListEntry> &sorted_list,
+                              std::vector<LambdaPair> *io_pairs) {
+    std::vector<LambdaPair> &pairs = *io_pairs;
+    std::vector<MAPStats> map_stats;
+    GetMAPStats(sorted_list, &map_stats);
+    for (auto & pair : pairs) {
+      pair.weight *=
+          GetLambdaMAP(sorted_list, pair.pos_index,
+                       pair.neg_index, &map_stats);
+    }
+  }
+};
+
+#if defined(__CUDACC__)
+// Helper functions
+
+// Labels of size 'n' are sorted in a descending order
+// If left is true,  find the number of elements > v; 0 if nothing is greater
+// If left is false, find the number of elements < v; 0 if nothing is lesser
+__device__ __forceinline__ int
+CountNumLabelsImpl(bool left, const float * __restrict__ labels, int n, float v) {
+  const float *labels_begin = labels;
+  int num_remaining = n;
+  const float *middle_item = nullptr;
+  int middle;
+  while (num_remaining > 0) {
+    middle_item = labels_begin;
+    middle = num_remaining / 2;
+    middle_item += middle;
+    if ((left && *middle_item > v) || (!left && !(v > *middle_item))) {
+      labels_begin = ++middle_item;
+      num_remaining -= middle + 1;
+    } else {
+      num_remaining = middle;
+    }
+  }
+
+  return left ? labels_begin - labels : labels + n - labels_begin;
+}
+
+__device__ __forceinline__ int
+CountNumLabelsToTheLeftOf(const float * __restrict__ labels, int n, float v) {
+  return CountNumLabelsImpl(true, labels, n, v);
+}
+
+__device__ __forceinline__ int
+CountNumLabelsToTheRightOf(const float * __restrict__ labels, int n, float v) {
+  return CountNumLabelsImpl(false, labels, n, v);
+}
+
+class SortedLabelList {
+ private:
+  // Labels sorted within the group
+  dh::caching_device_vector<float> dsorted_labels_;
+
+  // Original position of the labels before they are sorted descendingly within its groups
+  dh::caching_device_vector<uint32_t> doriginal_pos_;
+
+  // Segments within the original list that delineates the different groups
+  dh::caching_device_vector<uint32_t> group_segments_;
+
+  // Need this on the device as it is used in the kernels
+  dh::caching_device_vector<uint32_t> dgroups_;       // Group information on device
+
+  int device_id_{-1};                                 // GPU device ID
```

Comment:
I think you should get this value from generic parameters.

Reply:
it has been retrieved from generic param and passed here.